discovery-provider/src/tasks/index_user_bank.py

```diff
@@ -203,10 +226,255 @@ def process_create_userbank_instruction(
+    # Validate the amounts sent and received are equal
+    # TODO: Change this when payment router is going to take a cut
+    if (
+        balance_changes[sender_account]["change"]
+        + balance_changes[receiver_account]["change"]
+        != 0
+    ):
+        logger.error(
+            f"index_user_bank.py | ERROR: Sent and received amounts don't match. Sent = {balance_changes[sender_account]['change']}, Received = {balance_changes[receiver_account]['change']}"
+        )
+        return False
+    if purchase_metadata["price"] + balance_changes[sender_account]["change"] > 0:
+        logger.error(
+            f"index_user_bank.py | Purchase price exceeds sent amount. Sent={balance_changes[sender_account]['change']} Price={purchase_metadata['price']} | purchase_metadata={purchase_metadata} buyer={sender_account} seller={receiver_account}"
+        )
+        return False
+    return True
+
+
+def index_purchase(
+    session: Session,
+    receiver_user_id: Union[int, None],
+    receiver_account: str,
+    sender_user_id: int,
+    sender_account: str,
+    balance_changes: dict[str, BalanceChange],
+    purchase_metadata: PurchaseMetadataDict,
+    slot: int,
+    timestamp: datetime.datetime,
+    tx_sig: str,
+):
+    usdc_purchase = USDCPurchase(
+        slot=slot,
+        signature=tx_sig,
+        seller_user_id=receiver_user_id,
+        buyer_user_id=sender_user_id,
+        amount=purchase_metadata["price"],
+        content_type=purchase_metadata["type"],
+        content_id=purchase_metadata["id"],
+    )
+    logger.debug(
+        f"index_user_bank.py | Creating usdc_purchase for purchase {usdc_purchase}"
+    )
+    session.add(usdc_purchase)
+
+    usdc_tx_sent = USDCTransactionsHistory(
+        user_bank=sender_account,
+        slot=slot,
+        signature=tx_sig,
+        transaction_type=USDCTransactionType.purchase_content,
+        method=USDCTransactionMethod.send,
+        transaction_created_at=timestamp,
+        change=Decimal(balance_changes[sender_account]["change"]),
+        balance=Decimal(balance_changes[sender_account]["post_balance"]),
+    )
+    logger.debug(
+        f"index_user_bank.py | Creating usdc_tx_history send tx for purchase {usdc_tx_sent}"
+    )
+    session.add(usdc_tx_sent)
+    usdc_tx_received = USDCTransactionsHistory(
+        user_bank=receiver_account,
+        slot=slot,
+        signature=tx_sig,
+        transaction_type=USDCTransactionType.purchase_content,
+        method=USDCTransactionMethod.receive,
+        transaction_created_at=timestamp,
+        change=Decimal(balance_changes[receiver_account]["change"]),
+        balance=Decimal(balance_changes[receiver_account]["post_balance"]),
+    )
+    session.add(usdc_tx_received)
+    logger.debug(
+        f"index_user_bank.py | Creating usdc_tx_history received tx for purchase {usdc_tx_received}"
+    )
+
+
+def validate_and_index_purchase(
```

Comment:
nit: let's include USDC somewhere in the fn name here, and would be nice to reflect the fact that this fn can also save out a transaction rather than a purchase in the name